packages/@react-aria/landmark/src/useLandmark.ts

```diff
@@ -0,0 +1,245 @@
+/*
+ * Copyright 2022 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import {AriaLabelingProps} from '@react-types/shared';
+import {getFocusableTreeWalker} from '@react-aria/focus';
+import {HTMLAttributes, MutableRefObject, useEffect} from 'react';
+import {useLayoutEffect} from '@react-aria/utils';
+
+export type AriaLandmarkRole = 'main' | 'region' | 'search' | 'navigation' | 'form' | 'banner' | 'contentinfo' | 'complementary';
+
+export interface AriaLandmarkProps extends AriaLabelingProps {
+  role: AriaLandmarkRole
+}
+
+interface LandmarkAria {
+  landmarkProps: HTMLAttributes<HTMLElement>
+}
+
+type Landmark = {
+  ref: MutableRefObject<HTMLElement>,
+  role: AriaLandmarkRole,
+  label?: string,
+  lastFocused?: HTMLElement
+};
+
+class LandmarkManager {
+  private landmarks: Array<Landmark> = [];
+  private static instance: LandmarkManager;
+
+  private constructor() {}
+
+  public static getInstance(): LandmarkManager {
+    if (!LandmarkManager.instance) {
+      LandmarkManager.instance = new LandmarkManager();
+    }
+
+    return LandmarkManager.instance;
+  }
+
+  /**
+   * Return set of landmarks with a specific role.
+   */
+  public getLandmarksByRole(role: AriaLandmarkRole) {
+    return new Set(this.landmarks.filter(l => l.role === role));
+  }
+
+  public addLandmark({ref, role, label}: Landmark) {
+    if (!this.landmarks.find(landmark => landmark.ref === ref)) {
+
+       // TODO: Nested case?
+       // Inside should be after (DFS). So maybe we need to accept 4 or 8?
```

Comment:
Is this comment still valid?

Reply:
I've updated this to handle that case. I this that case would occur if someone added a landmark around an existing landmark. We hadn't run into that because 'usually' they are added in DOM order (outside first). Would love a second look at that logic though.

Reply:
I'll add a test case for this